File: examples/cpp/talker.cc
#include <chrono>
#include <string>
#include <thread>

#include "talosos/logging.h"
#include "talosos/messages.h"
#include "talosos/node.h"

int main(int argc, char** argv) {
  talos::Init(argc, argv);

  auto node = talos::Node::Create("talker");
  auto pub = node->Advertise<talos::msgs::String>("chatter");

  int count = 0;
  using namespace std::chrono_literals;
  while (talos::Ok()) {
    talos::msgs::String msg;
    msg.data = "hello from talker #" + std::to_string(count++);

    pub.Publish(msg);
    TALOS_INFO("publish '%s' -> %s", msg.data.c_str(), pub.key().c_str());

    std::this_thread::sleep_for(500ms);
  }

  TALOS_INFO("talker shutting down");
  return 0;
}

File: examples/python/talker.py
#!/usr/bin/env python3
"""Minimal Python publisher — mirror of examples/cpp/talker.cc.

Run with:
    source /opt/talosos/setup.bash
    python3 examples/python/talker.py
"""

from __future__ import annotations

import time

from talosos.messages import String
from talosos.runtime import Node, init, ok


def main() -> int:
    init()
    node = Node.create("py_talker")
    pub = node.advertise("chatter", String)   # relative name -> py_talker/chatter

    i = 0
    while ok():
        msg = String(data=f"hello from py_talker #{i}")
        pub.publish(msg)
        print(f"publish '{msg.data}' -> {pub.key}")
        i += 1
        time.sleep(0.5)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
